Include/internal/pycore_backoff.h

```diff
@@ -129,6 +134,14 @@ initial_unreachable_backoff_counter(void)
     return make_backoff_counter(0, UNREACHABLE_BACKOFF);
 }
 
+// Required to not get stuck in infinite specialization loops due to specialization failure.
+// We use 2 here as there are a few scenarios:
+// 1. Freshly specialized from unspecialized, in which case the counter will be 1.
+// 2. Re-specialized from deopt, in which case the counter will be 1.
+// 3. Deopt -> Specialize -> Deopt -> Specialize, in which case the counter will be 2.
+// We do not want the 3rd case.
+#define MAX_SPECIALIZATION_TRIES 2
```

Comment:
Is this necessary?

If we set the backoff counter to zero in `record_previous_instruction`, then:
* If specialization fails, we get re-specialization, but only once (no loop)
* If it succeeds, then tier 1 will re-specialization the next time the guard fails, which is what we want if you have just jitted the code.

It is also simple and fast.

Reply:
Let's fix those, rather than work around them.

Reply:
Why is this an infinite loop?

I would expect the following sequence:
* `STORE_SUBSCR_LIST_INT` [tracing]
* `record_previous_inst`
* set `counter = 0`
* `STORE_SUBSCR_LIST_INT` [non-tracing]
* `DEOPT`
* `STORE_SUBSCR` [non-tracing] (with counter == 0)
* Specialize `STORE_SUBSCR` (sets `counter = NON_ZERO`)
* `STORE_SUBSCR_LIST_INT` [non-tracing]
* `DEOPT`
* `STORE_SUBSCR` [non-tracing] (with counter != 0) 
next instruction [tracing]

Are you going to back to tracing before the instruction completes?

Reply:
> I think you need to dispatch using the non-tracing table, regardless of whether tracing is on or off, after specialization.

How do you detect that?